docs/source/custom-collections.rst

```diff
@@ -126,8 +154,227 @@ interface. Note that there is no required base class:
           see :ref:`deterministic-hashing`.
 
 
-Adding common dask methods to your class
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+.. _core-method-internals:
+
+Internals of the Core Dask Methods
+----------------------------------
+
+Dask has a few *core* functions (and corresponding methods) that implement
+common operations:
+
+- ``compute``: convert one or more dask collections into their in-memory
+  counterparts
+- ``persist``: convert one or more dask collections into equivalent dask
+  collections with their results already computed and cached in memory.
+- ``visualize``: given one or more dask collections, draw out the graph that
+  would be passed to the scheduler during a call to ``compute`` or ``persist``
+
+Here we briefly describe the internals of these functions to illustrate how
+they relate to the above interface.
+
+Compute
+~~~~~~~
+
+The operation of ``compute`` can be broken into three stages:
+
+1. **Graph Merging & Optimization**
+
+   First the individual collections are converted to a single large graph and
+   nested list of keys. How this happens depends on the value of the
+   ``optimize_graph`` keyword, which each function takes:
+
+   - If ``optimize_graph`` is ``True`` (default) then the collections are first
+     grouped by their ``__dask_optimize__`` methods. All collections with the
+     same ``__dask_optimize__`` method have their graphs merged and keys
+     concatenated, and then a single call to ``__dask_optimize__`` is made with
```

Comment:
If there are multiple types of collections, which `__dask_optimize__` method is called once? An arbitrary one?

Reply:
I also don't see this single call in the pseudo-code implementation.

Reply:
I believe that we group our graphs the optimization function, merge each collection of graphs and call the optimize function on those graphs together.  This happens separately for each optimization function.

Reply:
Matt is correct. I've updated this section to hopefully clarify this.